packages/webapp/src/containers/Finances/ExpenseDetail/index.jsx

```diff
@@ -1,222 +1,103 @@
-      if (targetDate === expenseDate) {
-        let id = e.expense_type_id;
-        total += parseFloat(e.value);
-        filteredExpenses.push(e);
-        if (!dict.hasOwnProperty(id)) {
-          dict[id] = {
-            type_name: this.getExpenseType(id),
-            items: [
-              {
-                note: e.note,
-                value: e.value,
-              },
-            ],
-          };
-        } else {
-          dict[id].items.push({
-            note: e.note,
-            value: e.value,
-          });
-        }
-      }
+import { expenseSelector, allExpenseTypeTileContentsSelector } from '../selectors';
+import { deleteExpense } from '../actions';
+import { useTranslation } from 'react-i18next';
+import { useSelector, useDispatch } from 'react-redux';
+import PureExpenseDetail from '../../../components/Finances/PureExpenseDetail';
+import { setPersistedPaths } from '../../hooks/useHookFormPersist/hookFormPersistSlice';
+import useHookFormPersist from '../../hooks/useHookFormPersist';
+import { updateExpense } from '../saga';
+
+const ExpenseDetail = ({ history, match }) => {
+  const { t } = useTranslation();
+  const dispatch = useDispatch();
+
+  useHookFormPersist(); // To clear form history after editing
+
+  const isEditing = match.path.endsWith('/edit');
+
+  const { expense_id } = match.params;
+
+  const allSortedExpenseTypes = useSelector(allExpenseTypeTileContentsSelector);
+  const expenses = useSelector(expenseSelector);
+
+  const expense = expenses.find((record) => record.farm_expense_id === expense_id);
+
+  useEffect(() => {
+    if (!expense) {
+      history.replace('/unknown_record');
     }
+  }, [expense, history]);
 
-    this.setState({
-      expenseItems: Object.values(dict),
-      total: total.toFixed(2),
-      filteredExpenses,
-    });
-  }
-
-  getExpenseType(id) {
-    const { expenseTypes } = this.props;
-    for (let type of expenseTypes) {
-      if (type.expense_type_id === id) {
-        return this.props.t(`expense:${type.expense_translation_key}`);
+  const expenseTypeReactSelectOptions = allSortedExpenseTypes
+    // Retired at end
+    .sort((a, b) => {
+      if (a.deleted && !b.deleted) {
+        return 1;
+      } else if (b.deleted && !a.deleted) {
+        return -1;
+      } else {
+        return 0;
       }
-    }
-    return 'TYPE_NOT_FOUND';
-  }
+    })
+    .map((type) => {
+      if (type.deleted && type.expense_type_id !== expense?.expense_type_id) {
+        // Dropdown should include the current expense's type even if it has been retired
+        // No other retired types should be shown
+        return;
+      } else {
+        const retireSuffix = type.deleted ? ` ${t('EXPENSE.EDIT_EXPENSE.RETIRED')}` : '';
+
+        return {
+          value: type.expense_type_id,
+          label: type.farm_id
+            ? type.expense_name + retireSuffix
+            : t(`expense:${type.expense_translation_key}`),
+        };
+      }
+    })
+    .filter(Boolean);
```

Comment:
I think we could do `expenseTypeTileContentsSelector` (sorry for the bad name) + `expenseTypeByIdSelector` (if the type is deleted one) and map...?

Reply:
My first thought was to use a map! But then got nervous because I've never actually used one before 😆 I will for sure try it!!

Reply:
Sorry for the lack of the explanation, the result of `sort()` and `filter()` is the same as not deleted types + the type of the current expense (if the type is retired):
```
[
   ...useSelector(expenseTypeTileContentsSelector),
   useSelector(expenseTypeByIdSelector(expense.expense_type_id))
]
```
correct?

So you can do:
```
theArrayAbove.map((type) => {
    const retireSuffix = type.deleted ? ` ${t('EXPENSE.EDIT_EXPENSE.RETIRED')}` : '';

     return {
       value: type.expense_type_id,
       label: type.farm_id
         ? type.expense_name + retireSuffix
        : t(`expense:${type.expense_translation_key}`),
    };
})
```
My intention was to remove the loops, sort and filter!

Reply:
Okay, I think that's about the same as what I figured (I was hoping to use a new datatype but I guess it's not necessary!) and yes it's WAY shorter!

But you would still need the deleted check right before appending to the array?

```js
  const expenseTypeReactSelectOptions = [
    ...sortedExpenseTypes,
    currentExpenseType.deleted && currentExpenseType,
  ].map((type) => {
    // as above
  });
```

Reply:
Mine was just an example so you can chain `map`!
Other than that, the code above looks great!